File: sdks/scala/src/main/scala/do_/mongo/Projections.scala
package do_.mongo

/**
 * MongoDB projection builders for controlling which fields are returned.
 */
object Projections:

  /**
   * Includes the specified fields.
   */
  def include(fields: String*): Bson =
    Bson(fields.map(_ -> 1).toMap)

  /**
   * Excludes the specified fields.
   */
  def exclude(fields: String*): Bson =
    Bson(fields.map(_ -> 0).toMap)

  /**
   * Excludes the _id field.
   */
  def excludeId: Bson =
    Bson.doc("_id" -> 0)

  /**
   * Projects array elements matching a condition.
   */
  def elemMatch(field: String, filter: Bson): Bson =
    Bson.doc(field -> Map("$elemMatch" -> filter.toMap))

  /**
   * Projects the first matching array element.
   */
  def elemMatchFirst(field: String): Bson =
    Bson.doc(s"$field.$$" -> 1)

  /**
   * Projects a slice of an array field.
   */
  def slice(field: String, limit: Int): Bson =
    Bson.doc(field -> Map("$slice" -> limit))

  /**
   * Projects a slice of an array field with skip and limit.
   */
  def slice(field: String, skip: Int, limit: Int): Bson =
    Bson.doc(field -> Map("$slice" -> List(skip, limit)))

  /**
   * Projects metadata (e.g., text score).
   */
  def metaTextScore(field: String): Bson =
    Bson.doc(field -> Map("$meta" -> "textScore"))

  /**
   * Combines multiple projections.
   */
  def fields(projections: Bson*): Bson =
    projections.foldLeft(Bson.empty)(_ ++ _)

  /**
   * Creates a computed field projection.
   */
  def computed(field: String, expression: Any): Bson =
    Bson.doc(field -> expression)

  /**
   * Extension methods for fluent projection building.
   */
  extension (projection: Bson)
    def and(other: Bson): Bson = fields(projection, other)

File: sdks/kotlin/src/main/kotlin/do_/mongo/Projections.kt
package do_.mongo

/**
 * Projection builders for MongoDB queries.
 *
 * Example usage:
 * ```kotlin
 * import do_.mongo.Projections.*
 *
 * // Include specific fields
 * val projection = include("name", "email")
 *
 * // Exclude fields
 * val projection = exclude("password", "internal")
 *
 * // Combine projections
 * val projection = fields(
 *     include("name", "email"),
 *     excludeId()
 * )
 * ```
 */
object Projections {

    /**
     * Creates a projection that includes the specified fields.
     */
    fun include(vararg fields: String): Bson =
        Document(fields.associateWith { 1 })

    /**
     * Creates a projection that excludes the specified fields.
     */
    fun exclude(vararg fields: String): Bson =
        Document(fields.associateWith { 0 })

    /**
     * Creates a projection that excludes the _id field.
     */
    fun excludeId(): Bson = Document("_id", 0)

    /**
     * Creates a projection for the first matching array element using $.
     */
    fun elemMatch(field: String, filter: Bson): Bson =
        Document(field, Document("\$elemMatch", filter.toBsonDocument()))

    /**
     * Creates a projection for text search score.
     */
    fun metaTextScore(field: String): Bson =
        Document(field, Document("\$meta", "textScore"))

    /**
     * Creates a slice projection for arrays.
     */
    fun slice(field: String, limit: Int): Bson =
        Document(field, Document("\$slice", limit))

    /**
     * Creates a slice projection with skip and limit.
     */
    fun slice(field: String, skip: Int, limit: Int): Bson =
        Document(field, Document("\$slice", listOf(skip, limit)))

    /**
     * Combines multiple projections.
     */
    fun fields(vararg projections: Bson): Bson {
        val combined = Document()
        for (projection in projections) {
            combined.putAll(projection.toBsonDocument())
        }
        return combined
    }

    /**
     * Combines a collection of projections.
     */
    fun fields(projections: Collection<Bson>): Bson = fields(*projections.toTypedArray())

    /**
     * Creates a computed field projection.
     */
    fun computed(field: String, expression: Any?): Bson =
        Document(field, expression)
}
